test/FunctionalTests/CompressionTests.cs

```diff
@@ -0,0 +1,311 @@
+            }
+
+            // Arrange
+            var url = Fixture.DynamicGrpc.AddUnaryMethod<object, HelloRequest, HelloReply>(WriteMessageWithoutCompression);
+
+            var requestMessage = new HelloRequest
+            {
+                Name = "World"
+            };
+
+            var requestStream = new MemoryStream();
+            MessageHelpers.WriteMessage(requestStream, requestMessage, "gzip");
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, url);
+            httpRequest.Headers.Add(GrpcProtocolConstants.MessageEncodingHeader, "gzip");
+            httpRequest.Headers.Add(GrpcProtocolConstants.MessageAcceptEncodingHeader, "gzip");
+            httpRequest.Content = new GrpcStreamContent(requestStream);
+
+            // Act
+            var response = await Fixture.Client.SendAsync(httpRequest).DefaultTimeout();
+
+            // Assert
+            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+
+            // The overall encoding is gzip but the actual response does not use compression
+            Assert.AreEqual("gzip", response.Headers.GetValues("grpc-encoding").Single());
+
+            var responseMessage = MessageHelpers.AssertReadMessage<HelloReply>(await response.Content.ReadAsByteArrayAsync().DefaultTimeout());
+            Assert.AreEqual("Hello World", responseMessage.Message);
+
+            Assert.AreEqual(StatusCode.OK.ToTrailerString(), Fixture.TrailersContainer.Trailers[GrpcProtocolConstants.StatusTrailer].ToString());
+        }
+
+        [Test]
+        public async Task SendUncompressedMessageToServiceWithCompression_ResponseCompressed()
+        {
+            // Arrange
+            var requestMessage = new HelloRequest
+            {
+                Name = "World"
+            };
+
+            var requestStream = new MemoryStream();
+            MessageHelpers.WriteMessage(requestStream, requestMessage);
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, "Compression.CompressionService/SayHello");
+            httpRequest.Content = new GrpcStreamContent(requestStream);
+
+            // Act
+            var response = await Fixture.Client.SendAsync(httpRequest).DefaultTimeout();
+
+            // Assert
+            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+            Assert.AreEqual("gzip", response.Headers.GetValues("grpc-encoding").Single());
+
+            var responseMessage = MessageHelpers.AssertReadMessage<HelloReply>(await response.Content.ReadAsByteArrayAsync().DefaultTimeout(), "gzip");
+            Assert.AreEqual("Hello World", responseMessage.Message);
+
+            Assert.AreEqual(StatusCode.OK.ToTrailerString(), Fixture.TrailersContainer.Trailers[GrpcProtocolConstants.StatusTrailer].ToString());
+        }
+
+        [Test]
+        public async Task SendIdentityGrpcEncodingToServiceWithCompression_ResponseUncompressed()
```

Comment:
the name of the test case seems a bit off, as the reason for responding with "identity" is the grpc-accept-encoding header only containing "identity", not related to what the request compression was.

Reply:
Renamed to SendIdentityGrpcAcceptEncodingToServiceWithCompression_ResponseUncompressed